### DISCUSSION TITLE
Causal Loop Diagrams (CLD)

### DISCUSSION THREAD

**Discussion by bbak** [mermaid-js/mermaid#4308] [Ideas]
I would be interested in building Causal Loop Diagrams (CLD) using Mermaid.js. I [already tried using the Flowchart type](https://kroki.io/mermaid/svg/eNp9lMFO20AQhu9-ihFSUCuBRZz0wqFSIIgiVNESBIc1hyUe26usx-nuOqhA34bH6C0v1tlNCXYSkSi2bM8_88834_R6z4qUO4Zn2Mt1_TgtpXF74ZLqDCdzOVVU8I3B0QHsGUmz91tfjuCP__Z60VoKN-Moss1DYeS8hAvK1EJlTVNF53E_7osJ6gfrFmisW75SRsoep2SkUzVJjQZGBZK7j4A_h4dfXyYvcB4nQZvs1pZ1iXBZV3N0SE87lIkYEcmyQg6-oLw2VahmfzVSK07kIGOpqZgCGpfSN0mZbqh4D7iHbtYhZx2Iu4bstEw597SElbVHZWZWVjNUOzVDoTyOGDrSG2Zq59J49xAhZT6yLR7F_Sjig7hFoyUxIPDClFrKlBRxHwauTBF3a4-YAv8-iXbzFXKe-8_duEHch_1wTv6fB1G4K0ZN_shYIF_-NSld1uXy1XRoX3n9MEQn62iDGheSHHbAQ84UkFomW-LBmzglXwtGM_c2hYOU4InJajXliY9_k6zUbFeWoZgoIpCNhXbZ_ZRuW6sTaE0Up_dPu8hO_KoyghNGF4ULcY3T0vnyqvDjhcZb_LleoQUXOCPHU0Xedj-j7YxJSJWI7549PCnU3GLJHL2Tt63bFI6DLBGFRt-3J8nLOjfogu_N8FMeWXTKfrmIN0MfLobv8TRO1uckClfiR81tbcL1qUPQqoGUGu-EiyjMwkzMyhxt-VqLebpaFcTL54KnG0a34rndx5Djh2JeW3ayQDjLc5z5TfK6E6yYHNPaYsWNjP3rwgdxJ0tDWFYcdwCZtDblvxiNForGgV6-NrnbWOA-65K2LqWPXuyr8GL_A4jRr-o), but the positioning of the nodes is obviously optimized for a flow and not (feedback-)loops. Nevertheless, the flowchart type could serve as a base and the flowchart syntax may need just minor changes (with a little bit of luck even no changes at all).

What should be changed compared to a Flowchart is the following:

- Edges should always be shaped like a section of a circle (or loop).
- Positioning of nodes doesn't really have a direction like TD or similar; it could be more like _inside-out_ or like _Nodes at the top of the sourcecode have a tendency towards the center_
- Any edge must denote the type of relationship between two nodes: either via a label like S, O or +, - to indicate if the edge represents a positive (**S**ame direction) or negative (**O**pposite direction) relationship or by utilizing different edge styles.
- Ability to label loops: a variant of subsections may suit that.
- Ability to indicate a _delay_ on an edge.

The positioning of the nodes, and therefore length of the edges is quite important since the latter is an indicator for the duration of the (feedback-)loop.

Here are some resources regarding the construction of CLDs: 

- [Guidelines for Drawing Causal Loop Diagrams](https://www.cs.toronto.edu/~sme/SystemsThinking/GuidelinesforDrawingCausalLoopDiagrams.pdf)
- The chapter on CLD in [Systems Mapping - How to build and use causal models of systems](https://link.springer.com/book/10.1007/978-3-031-01919-7) (open access PDF or E-Pub from Springer) 
- A huge and quite famous example from the UK Government, the [Obesity System Map](https://assets.publishing.service.gov.uk/government/uploads/system/uploads/attachment_data/file/296290/obesity-map-full-hi-res.pdf). That one uses different edge styles to distinguish S/O relationships. Additionally, I assume due to the size of the map, they didn't label the loops.

**Comment by sidharthv96** [comment]
@bbak can you please create an issue for this?
We have an issue template for new diagram suggestions. 

https://github.com/mermaid-js/mermaid/issues/new?assignees=&labels=Status%3A+Triage%2CType%3A+Enhancement%2CType%3A+New+Diagram&projects=&template=diagram_proposal.yml

**Reply by dudung** [reply]
Is there any progress on this? It is not continued to an issue as I saw it. I am also expecting this feature.

**Edit**
There is an issue about this #4453.

**Reply by bbak** [reply]
Yes, I created this one as requested. Apart from that, I'm not aware that something happened.

**Reply by dudung** [reply]
Is there any we can do? Should be it technical or we can discuss the syntaxes or anything else?

**Reply by timjore** [reply]
I need this as well. What kind of syntax might be helpful for creating CLDs in Mermaid?
